### DISCUSSION TITLE
Include other recipes

### DISCUSSION THREAD

**Discussion by gabevenberg** [cooklang/spec#55] [Ideas]
In my recipe database, I have a sauces folder. In many of my recipes, I call for one of the sauces in my sauces folder.

It would be nice if the cooklang spec supported this sort of structure. If I am, say, cooking chicken enchilada casserole, which needs enchilada sauce, I can link the enchilada sauce recipe from the casserole recipe. Then, when I generate my shopping list, cooklang will scale the enchilada sauce recipe as necessary and add the needed ingredients to the shopping list.

**Comment by dubadub** [comment]
That's a valid use-case. Similar to `include` in other languages, right?

**Reply by jakeanq** [reply]
This does sound very useful.  It would also provide a way to do meal plans - just write a regular recipe which includes all the required dishes.

Perhaps if an ingredient name matches the name of a `.cook` file in the search path (maybe the current working dir plus subdirectories for example, although it may be best for this to allow implementation-defined additional paths), it is considered an include?  This has an advantage of not requiring any additional syntax.

For example if I put `@enchilada sauce{1 cup}` in `chicken_enchilada_casserole.cook`, I need only define a file `enchilada_sauce.cook` and it is automatically treated as an include and linked.  The unit specified could then link to the servings in the included recipe for scaling.

I would define filename matching as being case insensitive with underscores, dashes and spaces being treated as the same character, for example `chicken casserole` matches `chicken casserole.cook`, Chicken_Casserole.cook, `chicken-casserole.COOK` but not `chickencasserole.cook`.  I would also define the presence of two recipes with the same name and different case or different 'space' characters as being an error.

**Reply by jakeanq** [reply]
One drawback is this requires recipe names to be globally-unique, so perhaps some sort of support for folder structures and/or namespacing should be used - if all the recipes come from one author I don't think this would be an issue, but if you imported a bunch of recipes from a different source that would have the potential to result in conflicts.

**Reply by gabevenberg** [reply]
Why not just support relative and absolute paths to the recipe file? No need to enforce globably unique names. Perhaps one could import ingredients from other files at the top of the file so one only has to maintain the filesystem path in one place.

**Reply by dubadub** [reply]
Agree about using just paths without global names.

One can reuse recipes and include them as a file relative to the recipe file. The number shows how much of default servings to use from the recipe.

```
Add @../Sauces/Guacamole.cook{1/10} and mix well.
```

It can be optional

```
Top with @?../Sauces/Enchilada sauce.cook{}.
```

If not specified default is 1.

It can be one of variants:

```
Add @pesto sauce{100%g} | @../Sauces/Pesto sauce.cook{1/3} and mix well.
```

Scaling works the same way, if non-linear scaling required one can specify

```
>> servings: 2|4|8
Add @../Sauces/Guacamole.cook{1/10|1/4|1/2} and mix well.
```

**Reply by briansunter** [reply]
My preference would be to just have hyperlinks and mention the referenced recipe in the ingredients, rather than something like a import.

**Reply by ziofat** [reply]
I don't think scaling with fraction is a good idea. The recipes I am willing to include are always sauces or stock, they are not designed for serving. A stock recipe may `yield` 1kg but will not serve 1. I'd like to directly include the recipe with specific unit and amount.

**Reply by jakeanq** [reply]
@gabevenberg / @dubadub - I think you're right, full or relative paths probably does make the most sense.

@briansunter Perhaps an option for whether to include sub-recipes on the list generation?  I would envisage it being displayed as a hyperlink in the recipe proper but I could also see potential uses (for example the above-mentioned meal planner idea) for including all of the required recipes in the list.

@ziofat I would imagine if you include it unitless (or maybe there's a specific "servings" unit) it would be connected to the servings, but if you specify other units it will leave it.

**Reply by briansunter** [reply]
You could also add a list of sub recipes in a tag as well, and the meal prep app could include those related recipes without a special syntax in cooklang (canonical metadata for subrecipes).
```
>>name: Chicken Enchiladas
>>subrecipes: Enchilada Sauce, Pickled Jalapeños
```

**Reply by dubadub** [reply]
@briansunter I like your solution in a way that it's very simple. But it will allow to facilitate simple UIs only. Having sub-recipes inside the main recipe adds context and makes UI aware that around this step a sub-recipe should be prepared. So it can be referenced inside step. Also sub-recipes sometimes can be purchased and used as a regular ingredient.

**Comment by dubadub** [comment]
Picking up that for development. Drafting the proposal here https://github.com/cooklang/spec/pull/118.

**Comment by dubadub** [comment]
The latest CLI now supports linking other recipes for reading and shopping lists as it proposed.
